### PR TITLE
Added Camera Clamping

### PR DESCRIPTION
**Pull request by JohnCodd**

### CODE REVIEW

**Review by rafael-plugge — CHANGES_REQUESTED** [review]
There is one change + one thing missing.

Change:
``` cpp
math::Vector2f clampTopLeft;
math::Vector2f clampSize;
bool clamp = false;
```
would recommend changing to
``` cpp
std::optional<math::Rectf> clamp;
```

and its missing the inner rectangle give away (there is a small rectangle that if the player is inside it, the camera doesn't follow it).

**Comment by JohnCodd** [comment]
Changes have been made

File: ArgoProject/src/parameters/CameraParameters.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+		CameraParameters(CameraParameters &&) = default;
+		CameraParameters & operator=(CameraParameters &&) = default;
+
+	public: // Public Static Functions
+	public: // Public Member Functions
+	public: // Public Static Variables
+	public: // Public Member Variables
+		std::optional<app::Entity> targetEntity = std::nullopt;
+		std::optional<math::Rectf> clampRect;
+		std::optional<bool> clamp = false;
```
Comment (rafael-plugge) [inline ArgoProject/src/parameters/CameraParameters.h]: member variable not used

File: ArgoProject/src/components/Camera.h [hunk 2]
```diff
@@ -21,9 +21,11 @@ namespace app::comp
 	public: // Public Member Functions
 	public: // Public Static Variables
 	public: // Public Member Variables
-		math::Vector2f position;
+		math::Vector2f center;
 		math::Vector2f size;
 		std::optional<app::Entity> target;
+		std::optional<math::Rectf> clampRect;
+		std::optional<float> internalClampSize;
```
Comment (rafael-plugge) [inline ArgoProject/src/components/Camera.h]:
```suggestion
		std::optional<math::Vector2f> internalClamp;
```
Needs to have different values for x-axis vs y-axis.

File: ArgoProject/src/factories/scenes/LevelSceneFactory.cpp [hunk 3]
```diff
@@ -12,7 +12,7 @@ std::vector<app::Entity> app::fact::sce::LevelSceneFactory::create()
 	auto level = fact::LevelFactory().create();	
 	entities.insert(entities.end(), level.begin(), level.end());
 	entities.push_back(player);
-	entities.push_back(fact::CameraFactory(player).create());
+	entities.push_back(fact::CameraFactory(app::par::CameraParameters{ player, math::Rectf(app::math::Vector2f(-800, 500), app::math::Vector2f(2000, 1000)), true, 20 }).create());
```
Comment (rafael-plugge) [inline ArgoProject/src/factories/scenes/LevelSceneFactory.cpp]:
```suggestion
	auto cameraParams = app::par::CameraParameters();
	cameraParams.targetEntity = player;
	cameraParams.clampRect = math::Rectf({ -800.0f, 500.0f }, { 2000.0f, 1000.0f});
	cameraParams.internalClamp = math::Vector2f(20.0f, 60.0f); // for when this gets changed
	entities.push_back(fact::CameraFactory(std::move(cameraParams)).create());
```
makes it more readable.

File: ArgoProject/src/parameters/CameraParameters.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		std::optional<float> internalClampSize;
```
Comment (rafael-plugge) [inline ArgoProject/src/parameters/CameraParameters.h]:
```suggestion
		std::optional<math::Vector2f> internalClamp;
```
internal clamp needs varying x-axis and y-axis values.

File: ArgoProject/src/systems/CameraSystem.cpp [hunk 5]
```diff
@@ -22,7 +22,61 @@ void app::sys::CameraSystem::update(app::time::seconds const & dt)
 		if (camera.target.has_value())
 		{
 			auto const & targetLocation = targetView.get(camera.target.value());
-			camera.position = targetLocation.position - (camera.size / 2.0f);
+			math::Vector2f targetPosition = targetLocation.position;
+			if (camera.internalClampSize.has_value())
+			{
+				math::Vector2f clampPosition = camera.center;
```
Comment (rafael-plugge) [inline ArgoProject/src/systems/CameraSystem.cpp]: Not needed, apply changes to `camera.center` directly.

File: ArgoProject/src/systems/CameraSystem.cpp [hunk 6]
```diff
@@ -22,7 +22,61 @@ void app::sys::CameraSystem::update(app::time::seconds const & dt)
...
+				bool followPlayer = false;
```
Comment (rafael-plugge) [inline ArgoProject/src/systems/CameraSystem.cpp]: no need fi changes applied directly

File: ArgoProject/src/systems/CameraSystem.cpp [hunk 7]
```diff
@@ -22,7 +22,61 @@ void app::sys::CameraSystem::update(app::time::seconds const & dt)
...
+				if (followPlayer)
+				{
+					camera.center = clampPosition;
+				}
+			}
+			else
+			{
+				camera.center = targetPosition;
+			}
+			//camera.position = targetLocation.position - (camera.size / 2.0f);
```
Comment (rafael-plugge) [inline ArgoProject/src/systems/CameraSystem.cpp]: Unneeded comment

**Review by rafael-plugge — CHANGES_REQUESTED** [review]

File: ArgoProject/src/systems/CameraSystem.cpp [hunk 8]
```diff
@@ -22,7 +22,50 @@ void app::sys::CameraSystem::update(app::time::seconds const & dt)
 		if (camera.target.has_value())
 		{
 			auto const & targetLocation = targetView.get(camera.target.value());
-			camera.position = targetLocation.position - (camera.size / 2.0f);
+			math::Vector2f targetPosition = targetLocation.position;
```
Comment (rafael-plugge) [inline ArgoProject/src/systems/CameraSystem.cpp]:
```suggestion
			math::Vector2f const & targetPosition = targetLocation.position;
```
this way the compiler can optimise away the `targetPosition`

**Review by rafael-plugge — CHANGES_REQUESTED** [review]

**Review by rafael-plugge — APPROVED** [review]
:+1:

**Pull request merged**